### PR TITLE
refactored name generation code

### PR DESCRIPTION
**Pull request by taiyu-len**
Replaced the 2 classes in name.h with 4 functions in the `Name` namespace.

Name generation is done through `Name::get` for any kind of name, and `Name::generate` for human names,
which is pretty much the same as before other then a few lines that used `NameGenerator` directly.

the `load_names_from_file` used in init code has been moved to `Name::load_from_file`
the names are stored in a `names` map defined in name.cpp, replacing the `NameGenerator` singleton thingy.

the `std::bind` thing is no longer needed as a result of removing the singleton stuff.

use of `nameFlags` is more typesafe now, with `operator|` defined for it.
i removed some magic numbers from its definition to make it more clear.

the original comment for `uint32_tsFromFlags` described it incorrectly.
the new version `get_matching_groups` has a more accurate description

also added some tests for name generation.

### CODE REVIEW

**Comment by kevingranade** [comment]
Looks like a nice cleanup.  I'm wondering if we can make some assertions in the test about how the names are distributed, we periodically get complaints about the same names occurring over and over.

File: src/name.cpp [hunk 1]
```diff
@@ -1,164 +1,154 @@
...
 #include "name.h"
+#include "json.h"
 #include "string_formatter.h"
-#include "output.h"
 #include "translations.h"
 #include "rng.h"
 #include "cata_utility.h"
 
-NameGenerator::NameGenerator()
+static inline nameFlags operator&( nameFlags l, nameFlags r )
```
Comment (BevapDin) [inline src/name.cpp]: `static` *and* `inline` is a bit pointless. `static` means it only exists in this translation unit, but `inline` means it exists identical in several units. What is it?

File: src/name.cpp [hunk 2]
```diff
@@ -1,164 +1,154 @@
...
-        return std::string( _( "Tom" ) );
-    }
-
-    // Get number of matching names
-    int nNames = 0;
-    for( uint32_t type : types ) {
-        nNames += names[type].size();
-    }
+    auto matching_groups = get_matching_groups( searchFlags );
+    if( matching_groups.size() ) {
```
Comment (BevapDin) [inline src/name.cpp]: This should call `empty`.

File: src/name.cpp [hunk 3]
```diff
@@ -1,164 +1,154 @@
...
+static const str_flag gender_flags[] = {
+    { "male",   nameIsMaleName },
+    { "female", nameIsFemaleName },
+    { "unisex", nameIsUnisexName }
+};
+
+static nameFlags usage_flag( std::string const &usage )
+{
+    for( auto const &i : usage_flags ) {
+        if( usage == i.str ) {
```
Comment (Coolthulhu) [inline src/name.cpp]: `std::find( usage_flags.begin(), usage_flags.end(), usage )`
Though you'd need to change the C-style array to C++ `std::array` or `vector`, `map` etc.

**Pull request merged**